### DISCUSSION TITLE
Type inference of class attributes in ext. modules

### DISCUSSION THREAD

**Discussion by iyakushev** [microsoft/pyright#2973] [Q&A]
My pyright settings are listed below:
```toml
[tool.pyright]  # version 1.1.217
useLibraryCodeForTypes = true
reportMissingTypeStubs = false
typeCheckingMode = "strict"
``` 
I was following an official FastAPI guide on how to implement proper authentication and noticed that `form_data` attributes have type of `Unknown`
```python 
@auth.post("/token", response_model=Token)
async def login_for_access_token(form_data: OAuth2PasswordRequestForm = Depends()):
    reveal_type(form_data)  # Type of "form_data" is "OAuth2PasswordRequestForm"
    reveal_type(form_data.username)  # Type of "username" is "Unknown"
```
The definition of this class is following:
```python
# .venv/.../fastapi/security/oauth2

class OAuth2PasswordRequestForm:
    def __init__(
        self,
        grant_type: str = Form(None, regex="password"),
        username: str = Form(...),
        password: str = Form(...),
        scope: str = Form(""),
        client_id: Optional[str] = Form(None),
        client_secret: Optional[str] = Form(None),
    ):
        self.grant_type = grant_type
        self.username = username
        self.password = password
        self.scopes = scope.split()
        self.client_id = client_id
        self.client_secret = client_secret
```
At first I thought Pyright simply won't infer types from the signature given `strict` mode. But then I tried to simulate this issue (in the same module, mind you). Consider there is a class `Foo` with attribute `x` which lacks annotation:
```python
class Foo:
    def __init__(self, x: int) -> None:
        self.x = x

def foo() -> Foo:
    return Foo(11)

x = foo()
reveal_type(x.x)  # type of x is int
```
Pyright narrows its type to be `int` from the context of `__init__` signature.
Furthermore, If I manually put annotations on `OAuth2PasswordRequestForm` attributes inside `__init__` Pyright starts to infer types correctly. 
```python
# .venv/.../fastapi/security/oauth2

class OAuth2PasswordRequestForm:
    def __init__(...):
        self.username: str = username
        ...
        
# myproj/foo.py
@auth.post("/token", response_model=Token)
async def login_for_access_token(form_data: OAuth2PasswordRequestForm = Depends()):
    reveal_type(form_data)  # Type of "form_data" is "OAuth2PasswordRequestForm"
    reveal_type(form_data.username)  # Type of "username" is "str"
```
My question is: Is there a bug? Or I might be missing something?

**Comment by erictraut** [comment]
This is intended behavior. The `fastapi` library contains a "py.typed" marker file, which indicates that the library author has provided type declarations for the public interface of the library. But in this case, the author of `fastapi` has omitted annotations for some symbols, including the `username` instance variable in the `OAuth2PasswordRequestForm` class. Pyright will not infer types in "py.typed" libraries. The reasoning behind this is documented in [this guidance](https://github.com/python/typing/blob/master/docs/source/libraries.rst).

The CLI version of pyright has a command called "--verifytypes" that "py.typed" package authors can run on their library to find and fix missing type information. When I run "pyright --verifytypes fastapi" on the latest version of the library (0.73.0), I receive the following summary:

```
Symbols exported by "fastapi": 952
  With known type: 694
  With partially unknown type: 258

Type completeness score: 72.9%
```

If this library is important to you, please encourage the maintainer to improve the type completeness by adding missing type annotations.

In the meantime, you can work around the issue by manually deleting the "py.typed" marker file in your installed copy of `fastapi`. Pyright will then use its normal behaviors, which includes the use of type inference if `useTypesFromLibraryCode` is set to true.

**Reply by iyakushev** [reply]
Thank you for quick and detailed response!
